### PR TITLE
Generalized multi-level relaxation

### PR DESCRIPTION
**Pull request by AMLattanzi**
1. Building the mask array in **ERFFillPatcher.cpp** is a bit off a mess. The code works but it is hard to follow and not perfectly general.

2. The tradeoff between space and time interpolation has not been assessed. We currently do both for all the fill calls; this should be modified (currently inefficient).

3.  Grids to evolve has been removed. With the fast integrator, we will compute an update on all interior cells; but the set region is overwritten because we call apply_bcs() at the end of a fast iteration.

### CODE REVIEW

**Review by asalmgren — APPROVED** [review]

**Comment by AMLattanzi** [comment]
@WeiqunZhang can you focus on the ERF_FillPatcher.cpp/.H files? The new code for generating the mask array for set and relax regions is [here](https://github.com/erf-model/ERF/blob/e2999a6a7475a30a30ddb6b425316866b07421b3/Source/BoundaryConditions/ERF_FillPatcher.cpp#L115).

File: Source/BoundaryConditions/ERF_FillPatcher.H [hunk 1]
```diff
@@ -11,28 +13,42 @@ public:
...
+                 amrex::InterpBase* interp);
+
+    void BuildMask (amrex::BoxArray const& fba, int nghost, int nghost_set);
 
-    void registerCoarseData (amrex::Vector<amrex::MultiFab const*> const& crse_data,
+    void RegisterCoarseData (amrex::Vector<amrex::MultiFab const*> const& crse_data,
                              amrex::Vector<amrex::Real> const& crse_time);
 
+    void InterpFace (amrex::MultiFab& fine,
+                     amrex::MultiFab& crse,
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.H]:
```suggestion
                     amrex::MultiFab const& crse,
```

File: Source/BoundaryConditions/ERF_FillPatcher.H [hunk 2]
```diff
@@ -11,28 +13,42 @@ public:
...
+                     int mask_val);
+
+    void InterpCell (amrex::MultiFab& fine,
+                     amrex::MultiFab& crse,
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.H]:
```suggestion
                     amrex::MultiFab const& crse,
```

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 3]
```diff
@@ -162,14 +176,107 @@ void ERFFillPatcher::registerCoarseData (Vector<MultiFab const*> const& crse_dat
...
 
     m_crse_times[0] = crse_time[0]; // time of "old" coarse data
     m_crse_times[1] = crse_time[1]; // time of "new" coarse data
 
     m_dt_crse = crse_time[1] - crse_time[0];
 }
+
+
+void ERFFillPatcher::InterpFace (MultiFab& fine,
+                                 MultiFab& crse,
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]:
```suggestion
                                 MultiFab const& crse,
```

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 4]
```diff
@@ -162,14 +176,107 @@ void ERFFillPatcher::registerCoarseData (Vector<MultiFab const*> const& crse_dat
...
+          AMREX_HOST_DEVICE_PARALLEL_FOR_4D_FLAG(RunOn::Gpu,fbx,ncomp,i,j,k,n,
+          {
+              if (mask_arr(i,j,k) == mask_val) face_linear_interp_z(i,j,k,n,fine_arr,crse_arr,ratio);
+          });
+      } // IndexType::NODE
+  } // MFiter
+}
+
+void ERFFillPatcher::InterpCell (MultiFab& fine,
+                                 MultiFab& crse,
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]:
```suggestion
                                 MultiFab const& crse,
```

File: Source/BoundaryConditions/ERF_FillPatcher.H [hunk 5]
```diff
@@ -11,28 +13,42 @@ public:
...
+
     template <typename BC>
-    void fill (amrex::MultiFab& mf, amrex::Real time,
-              BC& cbc, amrex::Vector<amrex::BCRec> const& bcs, bool fill_subset=false);
+    void FillSet (amrex::MultiFab& mf, amrex::Real time,
+                  BC& cbc, amrex::Vector<amrex::BCRec> const& bcs);
+
+    template <typename BC>
+    void FillRelax (amrex::MultiFab& mf, amrex::Real time,
+                    BC& cbc, amrex::Vector<amrex::BCRec> const& bcs);
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.H:51]: The implementation of these two functions are almost identical, except for the mask. So maybe you can have a new private function that does the work, have `FillSet` and `FillRelax` call it with the right mask.

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 6]
```diff
@@ -62,87 +57,106 @@ void ERFFillPatcher::Define (BoxArray const& fba, DistributionMapping  fdm,
...
     }
 
-    // Box arrays for fine and coarse
-    BoxArray cf_fba(std::move(bl));
+    // Box arrays for the coarse data
     BoxArray cf_cba(std::move(cbl));
-    BoxArray cf_fba_s(std::move(bl_subset));
-    DistributionMapping cf_dm(cf_fba);
-    DistributionMapping cf_dm_s(cf_fba_s);
+    DistributionMapping cf_dm(cf_cba);
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]:
```suggestion
```
This is never used.

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 7]
```diff
@@ -62,87 +57,106 @@ void ERFFillPatcher::Define (BoxArray const& fba, DistributionMapping  fdm,
...
+
+    // Grow the complement boxes and trim with the bounding box
+    Vector<Box>& com_bl_v = com_bl.data();
+    for (int i(0); i<com_bl.size(); ++i) {
+        Box& bx = com_bl_v[i];
+        bx.grow(IntVect(-nghost,-nghost,0));
+        for (int idim(0); idim<AMREX_SPACEDIM; ++idim) {
+            if (bx.bigEnd(idim) > fba_bnd.bigEnd(idim)) bx.setBig(idim,fba_bnd.bigEnd(idim));
+            if (bx.smallEnd(idim) < fba_bnd.smallEnd(idim)) bx.setSmall(idim,fba_bnd.smallEnd(idim));
+        }
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]:
```suggestion
        bx &= fba_bnd;
```

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 8]
```diff
@@ -62,87 +57,106 @@ void ERFFillPatcher::Define (BoxArray const& fba, DistributionMapping  fdm,
...
+    }
+
+    // Do second complement with the grown boxes
+    com_ba.define(com_bl);
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]:
```suggestion
    com_ba.define(std::move(com_bl));
```

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 9]
```diff
@@ -62,87 +57,106 @@ void ERFFillPatcher::Define (BoxArray const& fba, DistributionMapping  fdm,
...
+    com_ba.complementIn(com_bl, fba_bnd);
+
+    // Fill mask based upon the com_bl BoxList
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp:143]: There is a more efficient way to do this. Since this pattern of setVal in a BoxArray's region can be useful in other places, I will add a function to amrex. After that is merged, we can do a new PR to use it.

File: Source/BoundaryConditions/ERF_FillPatcher.H (hunk 1 above)
Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.H]: addressed in commit [a471441](https://github.com/erf-model/ERF/pull/1233/commits/a47144147d2776e60249af76f4fee7ffb7448bff)

File: Source/BoundaryConditions/ERF_FillPatcher.H (hunk 2 above)
Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.H]: addressed in commit [a471441](https://github.com/erf-model/ERF/pull/1233/commits/a47144147d2776e60249af76f4fee7ffb7448bff)

File: Source/BoundaryConditions/ERF_FillPatcher.cpp (hunk 4 above)
Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.cpp]: addressed in commit [a471441](https://github.com/erf-model/ERF/pull/1233/commits/a47144147d2776e60249af76f4fee7ffb7448bff)

File: Source/BoundaryConditions/ERF_FillPatcher.cpp (hunk 3 above)
Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.cpp]: addressed in commit [a471441](https://github.com/erf-model/ERF/pull/1233/commits/a47144147d2776e60249af76f4fee7ffb7448bff)

File: Source/BoundaryConditions/ERF_FillPatcher.H (hunk 5 above)
Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.H:51]: Addressed in commit [a471441](https://github.com/erf-model/ERF/pull/1233/commits/a47144147d2776e60249af76f4fee7ffb7448bff)

File: Source/BoundaryConditions/ERF_FillPatcher.cpp [hunk 10]
```diff
@@ -58,91 +53,106 @@ ERFFillPatcher::ERFFillPatcher (BoxArray const& fba, DistributionMapping  fdm,
...
+    m_cf_mask->setVal(m_relax_mask);
 
-    // Two coarse patches to hold the data to be interpolated
-    m_cf_crse_data[0] = new MultiFab (cf_cba, cf_dm, m_ncomp, 0);
-    m_cf_crse_data[1] = new MultiFab (cf_cba, cf_dm, m_ncomp, 0);
+    // Populate mask array
+    if (nghost_set < 0) {
+        m_cf_mask->setVal(m_set_mask);
+        BuildMask(fba,nghost_set,m_set_mask-1);
+    }
```
Comment (WeiqunZhang) [inline Source/BoundaryConditions/ERF_FillPatcher.cpp]: Line 104 `m_cf_mask->setVal(m_relax_mask);` can be put into the else block of this `if` statement.

Comment (AMLattanzi) [reply Source/BoundaryConditions/ERF_FillPatcher.cpp]: Addressed in [2fe61a6](https://github.com/erf-model/ERF/pull/1233/commits/2fe61a6ad3ea0c1c1c09868d2f9cf4b13a7908ca)

**Review by WeiqunZhang — APPROVED** [review]
